lib/capabilitySets.js

```diff
@@ -0,0 +1,191 @@
+  const {capabilitySet: oldSet} = oldRecord;
+  const expectedSequence = oldSet.sequence + 1;
+  if(capabilitySet.sequence !== expectedSequence) {
+    throw new BedrockError(
+      'Could not update profile agent capability set; ' +
+      'unexpected sequence number.',
+      'InvalidStateError', {
+        public: true,
+        httpStatusCode: 409,
+        actual: capabilitySet.sequence,
+        expected: expectedSequence
+      });
+  }
+
+  // FIXME: Add new revocation code
+  // // determine which zcaps must be revoked
+  // const zcapsToRevoke = getZcapsToRevoke({oldSet, newSet: capabilitySet});
+
+  // // revoke old zcaps
+  // const profile = await profiles.get({id: profileId});
+  // const keystoreAgent = await profiles.getKeystoreAgent({id: profileId});
+  // await revokeCapabilities({profile, keystoreAgent, zcaps: zcapsToRevoke});
+
+  // update record
+  const query = {
+    'capabilitySet.profileAgent': profileAgentId,
+    'capabilitySet.sequence': oldSet.sequence
+  };
+  const $set = {
+    'meta.updated': Date.now(),
+    capabilitySet
+  };
+  const collection = utils.getCollection('profile-profileAgentCapabilitySet');
+  const result = await collection.update(query, {$set}, database.writeOptions);
+  if(result.result.n === 0) {
+    const details = {
+      profileAgent: profileAgentId,
+      httpStatusCode: 400,
+      public: true
+    };
+    throw new BedrockError(
+      'Could not update profile agent capability set; ' +
+      'set either not found or unexpected sequence number.',
+      'InvalidStateError', details);
+  }
+};
+
+// remove a capability set
+exports.remove = async ({profileAgentId}) => {
+  assert.string(profileAgentId, 'profileAgentId');
+
+  // update existing capability set to have no zcaps, revoking any as needed
+  const oldRecord = await exports.get({profileAgentId});
+  const {capabilitySet} = oldRecord;
+  capabilitySet.sequence++;
+  capabilitySet.zcaps = [];
+  await exports.update({capabilitySet});
+
+  // remove capability set if sequence matches
+  const query = {
+    'capabilitySet.profileAgent': profileAgentId,
+    'capability.sequence': capabilitySet.sequences
+  };
+  const collection = utils.getCollection('profile-profileAgentCapabilitySet');
+  const result = await collection.remove(query);
+  if(result.result.n === 0) {
+    const details = {
+      profileAgent: profileAgentId,
+      httpStatusCode: 404,
+      public: true
+    };
+    throw new BedrockError(
+      'Profile agent capability set with expected sequence not found.',
+      'NotFoundError', details);
+  }
+};
+
+function assertCapabilitySet(capabilitySet) {
+  assert.object(capabilitySet, 'capabilitySet');
+  assert.string(capabilitySet.profileAgent, 'capabilitySet.profileAgent');
+  assert.arrayOfObject(capabilitySet.zcaps, 'capabilitySet.zcaps');
+
+  const {sequence} = capabilitySet;
+  assert.number(sequence, 'capabilitySet.sequence');
+  if(!(sequence >= 0 && Number.isInteger(sequence))) {
```

Comment:
Seems like reversing the order of these would be more accurate.  Because javascript is weird: https://repl.it/repls/WateryPointlessSupercollider